content/blog/2024-01-26-more-secrets-options.md

```diff
@@ -0,0 +1,312 @@
+was the secrets handling in GitOps. So we worked out a scheme that stored
+secret material in a dedicated secrets store, using the External Secrets Operator
+to project that secret material into the applications using ESO's powerful and
+convenient abstraction and templating features. At that time, HahsiCorp Vault
+was well supported and popular in the community, as was using ESO to retrieve
+secrets from it. All of the major hyperscaler keystores are supported (AWS,
+Azure, GCP), but multi- or hybrid- cloud solutions that can be "self-contained"
+are either less well supported or the solutions themselves lean towards SaaS
+offerings.
+
+Almost two years later, the secrets landscape has shifted somewhat. As a Red
+Hat project, Validated Patterns initiative gravitates towards Red Hat-supported
+solutions, and neither HashiCorp Vault nor ESO are currently Red Hat supported.
+Meanwhile, the only backend we provided a code path to drive with ESO was Vault.
+
+This nullifies one of the major reasons we wanted to use ESO in the first place -
+namely, the ability to easily swap out secrets backends in case Vault was not
+usable for some reason. Earlier in 2023, one of our engineers did a proof of
+concept of ESO support using the AWS Secrets Manager backend - demonstrating that
+ESO delivered on its promise of multiple secret store support. Adapting ESO was
+the easy part - the hard part is building more abstraction into the VP secrets
+handling code that runs on pattern install.
+
+To this end, we decided we would expand secrets support by introducing at least
+one new backend - and we chose the kubernetes backend, because it is self-contained
+(that is, it can be run on-prem and requires no new products or projects to be
+installed), and is a useful vehicle for introducing an abstraction layer for
+validated patterns secret parsing. In addition, dealing with kubernetes secrets
+objects directly has the side effect of enabling us to provide a mechanism for
+users to inject their secrets directly into their patterns, bypassing the need
+to use any secrets manager or ESO at all. This also provides benefits to installations
+where only commercially supported solutions can be installed, since ESO is
+currently not commercially supported by any entity.
+
+In a nutshell, the new features depend on an abstraction of secret file parsing, so
+that the secrets are held in memory in a datastructure that is then processed and
+loaded by the appropriate backend code.
+
+Users of the pattern framework will be able to change secrets backends as straightforwardly
+as we can make possible. The only other change the user will need to make (to use another
+ESO backend) is to use the backend's mechanism to refer to keys. (For example: in Vault,
+keys have have names like `secret/data/global/config-demo`; in the Kubernetes backend
+it would just be the secret object name that's being used to store the secret material,
+such as `config-demo`).
```

Comment:
This seems like an opportunity (in the future) to split the key into a prefix and suffix.  
Vault advertising a default prefix of "secret/data/global/", k8s advertising "", and the chart supplying the "config-demo" suffix

Reply:
That *might* be a win on the complexity front; to me this is more of an argument to consolidate secrets handling; because spreading this convention over an arbitrary set of charts. Still, if we wind up adding an array of secret objects (making the structure open-ended), I don't feel like that would be a major improvement. How do you see the parameters working, if we were to do the prefix/suffix thing?